cpp/include/cudf/ast/linearizer.hpp

```diff
@@ -0,0 +1,259 @@
+
+  /**
+   * @brief Accepts a visitor class.
+   *
+   * @param visitor Visitor.
+   * @return cudf::size_type Index of device data reference for this instance.
+   */
+  cudf::size_type accept(detail::linearizer& visitor) const override;
+
+  const cudf::detail::fixed_width_scalar_device_view_base value;
+};
+
+/**
+ * @brief A node referring to data from a column in a table.
+ *
+ */
+class column_reference : public detail::node {
+  friend class detail::linearizer;
+
+ public:
+  /**
+   * @brief Construct a new column reference object
+   *
+   * @param column_index Index of this column in the table (provided when the node is
+   * evaluated).
+   * @param table_source Which table to use in cases with two tables (e.g. joins).
+   */
+  column_reference(cudf::size_type column_index,
+                   table_reference table_source = table_reference::LEFT)
+    : column_index(column_index), table_source(table_source)
+  {
+  }
+
+  /**
+   * @brief Get the column index.
+   *
+   * @return cudf::size_type
+   */
+  cudf::size_type get_column_index() const { return column_index; }
+
+  /**
+   * @brief Get the table source.
+   *
+   * @return table_reference
+   */
+  table_reference get_table_source() const { return table_source; }
+
+  /**
+   * @brief Get the data type.
+   *
+   * @param table Table used to determine types.
+   * @return cudf::data_type
+   */
+  cudf::data_type get_data_type(const table_view& table) const
+  {
+    return table.column(get_column_index()).type();
+  }
+
+  /**
+   * @brief Get the data type.
+   *
+   * @param left_table Left table used to determine types.
+   * @param right_table Right table used to determine types.
+   * @return cudf::data_type
+   */
+  cudf::data_type get_data_type(const table_view& left_table, const table_view& right_table) const
+  {
+    const auto table = [&] {
+      if (get_table_source() == table_reference::LEFT) {
+        return left_table;
+      } else if (get_table_source() == table_reference::RIGHT) {
+        return right_table;
+      } else {
+        CUDF_FAIL("Column reference data type cannot be determined from unknown table.");
+      }
+    }();
+    return table.column(get_column_index()).type();
+  }
+
+ private:
+  /**
+   * @brief Accepts a visitor class.
+   *
+   * @param visitor Visitor.
+   * @return cudf::size_type Index of device data reference for this instance.
+   */
+  cudf::size_type accept(detail::linearizer& visitor) const override;
+
+  cudf::size_type column_index;
+  table_reference table_source;
+};
+
+/**
+ * @brief An expression node holds an operator and zero or more operands.
+ *
+ */
+class expression : public detail::node {
+  friend class detail::linearizer;
+
+ public:
+  /**
+   * @brief Construct a new unary expression object.
+   *
+   * @param op Operator
+   * @param input Input node (operand)
+   */
+  expression(ast_operator op, node const& input) : op(op), operands({input})
```

Comment:
Is this still TODO?

Reply:
Yes, this is still an issue. It may be worth documenting that expression constructors cannot accept rvalue references.

Reply:
I've attempted a fix by deleting the r-value reference overloads here: https://github.com/bdice/cudf/pull/1
I feel it's simpler and produces better compiler error messages than using SFINAE.

Reply:
I agree deleting is better than SFINAE here. I commented on the PR - not sure if this actually worked when I tried it locally, but I didn’t have time to identify what was wrong. It might work fine and I could have made a mistake in testing.

Reply:
I've tried it locally, it works.

Reply:
Sweet. I can merge that fix into #5494 PR then. Thanks!

Reply:
Awesome! Glad I could help :)